backend/oasst_backend/utils/database_utils.py

```diff
@@ -46,17 +51,39 @@ def wrapped_f(self, *args, **kwargs):
                         if isinstance(result, SQLModel):
                             self.db.refresh(result)
                         return result
-                    except OperationalError:
-                        logger.info(f"Retry {i+1}/{num_retries} after possible DB concurrent update conflict.")
+                    except OperationalError as e:
+                        if e.orig is not None and isinstance(
+                            e.orig, (SerializationFailure, DeadlockDetected, UniqueViolation, ExclusionViolation)
+                        ):
+                            logger.info(f"{type(e.orig)} Inner {e.orig.pgcode} {type(e.orig.pgcode)}")
+                            if auto_commit != CommitMode.COMMIT:
+                                logger.info(
+                                    "Since it is a flush or rollback, let the outer call handle the transaction retry"
+                                )
+                                # TODO: Is returning result ok ?
+                                return result
+                        else:
+                            logger.info(f"Other kind of error {e}")
+                            OasstError(
+                                "DATABASE_OPERATION_ERROR",
+                                error_code=OasstErrorCode.DATABASE_OPERATION_ERROR,
+                                http_status_code=HTTPStatus.SERVICE_UNAVAILABLE,
+                            )
+                    except PendingRollbackError as e:
+                        logger.info(f"Retry {i+1}/{num_retries} after {e}")
                         self.db.rollback()
                 raise OasstError(
                     "DATABASE_MAX_RETIRES_EXHAUSTED",
                     error_code=OasstErrorCode.DATABASE_MAX_RETRIES_EXHAUSTED,
                     http_status_code=HTTPStatus.SERVICE_UNAVAILABLE,
                 )
             except Exception as e:
-                logger.error("DB Rollback Failure")
-                raise e
+                logger.error(f"DB Failure {type(e)} {e}")
```

Comment:
since we have here an additional catch-block maybe we should split the function into to branches .. one with retry (Commit-case) and the other without.